blockchain/upgrade.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2017 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package blockchain
+
+import (
+	"bytes"
+	"container/list"
+	"fmt"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/database"
+	"github.com/btcsuite/btcd/wire"
+)
+
+const (
+	// blockHdrOffset defines the offsets into a v1 block index row for the
+	// block header.
+	//
+	// The serialized block index row format is:
+	//   <blocklocation><blockheader>
+	blockHdrOffset = 12
+)
+
+// migrateBlockIndex migrates all block entries from the v1 block index bucket
+// to the v2 bucket. The v1 bucket stores all block entries keyed by block hash,
+// whereas the v2 bucket stores the exact same values, but keyed instead by
+// block height + hash.
+func migrateBlockIndex(db database.DB) error {
+	// Hardcoded bucket names so updates to the global values do not affect
+	// old upgrades.
+	v1BucketName := []byte("ffldb-blockidx")
+	v2BucketName := []byte("blockheaderidx")
+
+	err := db.Update(func(dbTx database.Tx) error {
+		v1BlockIdxBucket := dbTx.Metadata().Bucket(v1BucketName)
+		if v1BlockIdxBucket == nil {
+			return fmt.Errorf("Bucket %s does not exist", v1BucketName)
+		}
+
+		log.Info("Re-indexing block information in the database. This might take a while...")
+
+		v2BlockIdxBucket, err :=
+			dbTx.Metadata().CreateBucketIfNotExists(v2BucketName)
+		if err != nil {
+			return err
+		}
+
+		// Scan the old block index bucket and construct a mapping of each block
+		// to all child blocks.
+		childBlocksMap, err := readBlockTree(v1BlockIdxBucket)
+		if err != nil {
+			return err
+		}
+
+		// Use the block graph to calculate the height of each block.
+		blockHeights := determineBlockHeights(childBlocksMap)
+
+		// Now that we have heights for all blocks, scan the old block index
+		// bucket and insert all rows into the new one.
+		return v1BlockIdxBucket.ForEach(func(hashBytes, blockRow []byte) error {
```

Comment:
It's already pretty fast, so it probably doesn't matter, but it seems like this could be done in a single pass of reading the old bucket.  The records could be truncated while reading the block tree by saving the raw block header bytes and returning them along with the graph since it's already having to work with them to get the hash there anyways.   Then this could just range over the returned `blockHeights` map, look up the header bytes in the aforementioned map, and write the new entries.  Since it's all in a db transaction, any errors would roll everything back, so any failures wouldn't corrupt things.

I'm not sure if it's worth changing though since this is already done, works properly, and is only a one time operation.

Reply:
I did it this way to avoid keeping all headers in memory simultaneously as this migration runs.

Reply:
I though that might be the intent.  Of course they're all going to be loaded into memory shortly after anyways, but I have no strong objection to leaving this as it is.